---
layout: post
title: First results
description: ""
category: Thesis
tags: [thesis,first image]
---
{% include JB/setup %}

#### Important Note: If you get lost reading this post, you can go to the [Introduction post](http://kamynz.github.io/thesis/2015/07/30/Introduction-to-my-thesis/) to understant the objective of my thesis

##Creating an image with PCA results

This is the first attempt of creating an image with dimension reduction using PCA. However, the sum of the variance 
of the first three components is ~20%. 

## GETTING LIBRARIES
    
    library("seqinr")
    library("ade4")
    library("Biostrings")
    
## Function 1
    
    Processing_fragments<-function(PATH_FILES){
      
      #GETTING THE FILES AFTER FRAGMENTS OF 500
      files <-list.files(path=PATH_FILES, pattern=".fna500mer", full.names=T, recursive=FALSE)
      
      #GETTING THE FILES READING AS FASTA
      ncrna <-lapply(files, function(x) { read.fasta(x,seqonly = T) })
      
      
      fragmentsGeno1<-list()
      for(k in seq_along(ncrna[1]))
      {
        for(l in 1:10484)
        {
          fragmentsGeno1[l]<-ncrna[[k]][[l]]
          
        }
      }
      
      fragmentsGeno2<-list()
      for(k in seq_along(ncrna[2]))
      {
        for(l in 1:length(ncrna[[2]]))
        {
          fragmentsGeno2[l]<-ncrna[[k]][[l]]
          
        }
      }
    
      #GETTING ALL FRAGMENTS
      
      allFragments<-c(fragmentsGeno1,fragmentsGeno2)
    
      return(allFragments)
    
    }
    
## Function 2
    
    Getting_frequency_account<-function(allFragments,kmer){
      
      #CONVERTING LOS FRAGMENTOS DE CADA FILE A OBJETOS DE DNAString
      
      DNA_String_Set_list_ALL<-list()
      
      for(i in seq_along(allFragments))
      {
        DNA_String_Set_list_ALL[i]<-DNAStringSet(allFragments[[i]])
      }
      
      # counting oligonucleotide
      countGenome1_Tetra<-lapply(DNA_String_Set_list_ALL,function(x) {oligonucleotideFrequency((x),kmer, as.prob = T) })
      
      # MATRIX FOR THE PCA
      
      #names columns
      col_names<-dimnames(countGenome1_Tetra[[1]])
      col_names<-col_names[[2]]
      
      #names rows
      frag_names<-c(paste("frag",c(1:length(allFragments)),sep=""))
      
      #matrix for PCA
      matrix_PCA<-matrix(unlist(countGenome1_Tetra),nrow = length(allFragments),ncol=256,byrow = T,dimnames=list(frag_names,col_names))
      
      return(matrix_PCA)
      
    }
    
    
    # View(matrix_PCA)
    
## Function 3
    
    Getting_first_three_components<-function(matrix_for_PCA){
    
      ######## PCA with prcomp#########
      
      prcomp_All<-prcomp(matrix_for_PCA)
      
      #obtaing the sum of varianza of the first three components
      
      Var<-prcomp_All$sdev^2 / sum(prcomp_All$sdev^2)
      
      Varianza_3_first_comp<-Var[1:3]
      
      Varianza_3_first_comp_Porcent<-Varianza_3_first_comp*100
      
      Suma_total<-sum(Varianza_3_first_comp_Porcent)
      
      ## obteniendo eigen of first three components 
      
      loadings_prcomp<-prcomp_All$x
    
      #dim(loadings_prcomp)
      
      First_three_components<-loadings_prcomp[,c(1,2,3)]
      
      return(list("Components3"=First_three_components,"VarComponents3"=Suma_total))
    
    }
    
#Function 4
    
    Generating_hex_color_codes<-function(First_three_components){
      
      # getting min and max
      min<-min(First_three_components)
      max<-max(First_three_components)
      
      # getting ranges
      range_2_color<-c(min,max)
      range_RGB_color<-c(0,1)
      
      #making linear regression
      lm.out<-lm(range_RGB_color~range_2_color)
      
      #getting slope and intercept
      slope<-lm.out$coefficients[2]
      intercept<-lm.out$coefficients[1]
      
      #normalizing pca results to RGB
      new_Matriz<-(First_three_components*slope)+intercept
      
      #using funcion rgb to generate matrix of hex color code
      
      # list de cada r,g,b de cada fragmento
      
      Hex_Color_Vector<-vector()
      
      rgb_List_Each_Fragment<-list()
      
      row_Final<-length(new_Matriz[,1])
      
      columns_Final<-length(new_Matriz[1,])
      
      for(i in 1:row_Final){
        
        for(j in 1:columns_Final){
          
          red<-new_Matriz[i,1]
          green<-new_Matriz[i,2]
          blue<-new_Matriz[i,3]
          
          Hex_Color_Vector[i]<-rgb(red,green,blue,maxColorValue = 1)
          
        }
        
      }
      
      return(Hex_Color_Vector)
      
    }
    
## Calling all the funcionts in order
    
    allFragments<-Processing_fragments("/Users/CamilaMV/Desktop/TESIS/DATA")
    
    matrix_for_PCA<-Getting_frequency_account(allFragments,4)
    
    First_three_components<-Getting_first_three_components(matrix_for_PCA)
    
    Hex_color_vector<-Generating_hex_color_codes(First_three_components$Components3)
    
## Function
    
    #creating image
    
    Obtaining_dimensions_for_image<-function(Hex_color_Vector){
      
      len<-length(Hex_color_Vector)
      
      number_rows<-0
      number_cols<-0
      dimensions_vector<-vector()
      
      #first I have to know if len is odd or even
      
      if(len%%2==0){
        
        factors_of_hex_vector<-FUN(len)
        
        positive_factors<-factors_of_hex_vector$pos[factors_of_hex_vector$pos>=100]
        
        number_rows<-as.integer(positive_factors[1])
        
        number_cols<-as.integer(len/number_rows)
        
      }
      else{
        
        len<-len-1
        
        factors_of_hex_vector<-FUN(len)
        
        positive_factors<-factors_of_hex_vector$pos[factors_of_hex_vector$pos>=100]
        
        number_rows<-as.integer(positive_factors[1])
        
        number_cols<-as.integer(len/number_rows)
      }
      
      dimensions_vector<-c(number_rows,number_cols,len)
      
      return(list("nuRows"=number_rows,"nuCols"=number_cols,"lenForImage"=len))
    }
    
## Using function 5
    
    dim<-Obtaining_dimensions_for_image(Hex_color_vector)
    
    
## Using function 6 to generate the image after obtaining dimensions
    
    # CREANDO IMAGEN, SERA QUE TENGO QUE HACER ESCALA DE COLORES?
    
    ## test 3

    library(grid)
    
    Creating_image<-function(nu_rows,nu_cols,len_for_color_matriz,Hex_color_vector){
      
      grid_matrix<-matrix(unlist(Hex_color_vector[1:len_for_color_matriz]), 
                          nrow=nu_rows, ncol = nu_cols, byrow = TRUE)
      
      
      imageFinal<-grid.raster(grid_matrix, interpolate=FALSE)
      
      return(imageFinal)
      
    }
    
    FUN <- function(x) {
      x <- as.integer(x)
      div <- seq_len(abs(x))
      factors <- div[x %% div == 0L]
      factors <- list(neg = -factors, pos = factors)
      return(factors)
    }
    
## TESTING FUNCION 6 ### Notes => aparently is working
    
    Creating_image(dim$nuRows,dim$nuCols,dim$lenForImage,Hex_color_vector)
    
========================================================

## To-Do

1. `I have to normalize` the input data for the count of oligonucleotide frequency of Escheria coli. 
2. `I have to do a PCo and Non-metric Multidimensional Scaling with vegan library in R`.
3. `I have to construct a better function` but first I want to see what statistical techniques 
are the best to get significant biological variables.
4. `I still have to evaluate others lengths of fragments (e.g. 1000, 5000, etc)` and slide through
the genomes with overlaping windows`

========================================================
    
## Image of 10710 fragments of Ecoli, each fragment is 500bp long

![center](/Figs/Ecoli_Test1_PCA.png)

