---
layout: post
title: Second results
description: ""
category: Thesis
tags: [thesis, other images]
---
{% include JB/setup %}

### Progress during the first and second week

During these two weeks, I have been studying about multivariate statistics according to the GUide to STasitical Analysis in Microbial Ecology (GUSTA ME) [online guide](https://sites.google.com/site/mb3gustame/).

This guide helped to understand what R packages and data, I should use in order to get a better analysis. Also, I have improved the source code of different functions with the purpose of having better results.

### Function 1 => Processing_fragments(PATH_FILES)

  Processing_fragments<-function(PATH_FILES){
    
    #GETTING THE FILES AFTER FRAGMENTS OF 500
    files <-list.files(path=PATH_FILES, pattern=".fna500mer", full.names=T, recursive=FALSE)
    
    #GETTING THE FILES READING AS FASTA
    ncrna<-lapply(files, function(x) { read.fasta(x,as.string = T,seqonly = T) })
    
  
    ##GETTING ALL THE FRAGMENTS
    
    fragments_length_list<-vector()
    
    ## Loop to get the number of fragments for each file
    for(i in 1:length(files))
    {
      n<-length(ncrna[[i]])
      fragments_length_list[i]<-as.numeric(n)
      
    }
  
    ## APPENDING EACH TEMP LIST OF FRAGMENTS IF ALLFRAGMENTS
    
    allFragmentsBetter<-list()
    
    for(j in 1:length(ncrna))
    {
      inicio = 1
      final = fragments_length_list[j]
      
      #cat(inicio,"\n",final,"\n")
    
      temp_genome_list<-list()
      
      for(k in 1:final)
      {
        temp_genome_list[k]<-ncrna[[j]][[k]]     
      }
      
      allFragmentsBetter<-append(allFragmentsBetter,temp_genome_list)
      
    }
    
    return(allFragmentsBetter)
  }
  
### Function 2 => Getting_frequency_account(allFragments,kmer)

  Getting_frequency_account<-function(allFragments,kmer){
    
    #CONVERTING LOS FRAGMENTOS DE CADA FILE A OBJETOS DE DNAString
    
    DNA_String_Set_list_ALL<-list()
    
    for(i in seq_along(allFragments))
    {
      DNA_String_Set_list_ALL[i]<-DNAStringSet(allFragments[[i]])
    }
    
    # counting oligonucleotide
    countGenome1_Tetra<-lapply(DNA_String_Set_list_ALL,function(x) {oligonucleotideFrequency((x),kmer, as.prob = T) })
    
    # MATRIX FOR THE PCA
    
    #names columns
    col_names<-dimnames(countGenome1_Tetra[[1]])
    col_names<-col_names[[2]]
    
    #names rows
    frag_names<-c(paste("frag",c(1:length(allFragments)),sep=""))
    
    #matrix for PCA
    matrix_PCA<-matrix(unlist(countGenome1_Tetra),nrow = length(allFragments),ncol=256,byrow = T,dimnames=list(frag_names,col_names))
    
    return(matrix_PCA)
    
  }

### Function 3 => Getting_first_three_components(matrix_for_PCA_or_MDS)

  Getting_first_three_components<-function(matrix_for_PCA){
  
    ######## PCA with prcomp#########
    
    prcomp_All<-prcomp(matrix_for_PCA)
    
    #obtaing the sum of varianza of the first three components
    
    Var<-prcomp_All$sdev^2 / sum(prcomp_All$sdev^2)
    
    Varianza_3_first_comp<-Var[1:3]
    
    Varianza_3_first_comp_Porcent<-Varianza_3_first_comp*100
    
    Suma_total<-sum(Varianza_3_first_comp_Porcent)
    
    ## obteniendo eigen of first three components 
    
    loadings_prcomp<-prcomp_All$x
  
    #dim(loadings_prcomp)
    
    First_three_components<-loadings_prcomp[,c(1,2,3)]
    
    return(list("Components3"=First_three_components,"VarComponents3"=Suma_total))
  
  }
  
    Getting_first_three_components<-function(matrix_for_PCA){
    
      ######## PCA with prcomp#########
      
      prcomp_All<-prcomp(matrix_for_PCA)
      
      #obtaing the sum of varianza of the first three components
      
      Var<-prcomp_All$sdev^2 / sum(prcomp_All$sdev^2)
      
      Varianza_3_first_comp<-Var[1:3]
      
      Varianza_3_first_comp_Porcent<-Varianza_3_first_comp*100
      
      Suma_total<-sum(Varianza_3_first_comp_Porcent)
      
      ## obteniendo eigen of first three components 
      
      loadings_prcomp<-prcomp_All$x
    
      #dim(loadings_prcomp)
      
      First_three_components<-loadings_prcomp[,c(1,2,3)]
      
      return(list("Components3"=First_three_components,"VarComponents3"=Suma_total))
    
    }


