---
layout: post
title: Improvement on Third week
description: "New advances"
category: Thesis
tags: [thesis, new images]
---

{% include JB/setup %}

### Third Results

During this week, I have discussed my results with members of the BCEM [web page](http://bcem.uniandes.edu.co/wordpress/). Then, I decided to use three genomes (E. coli, Shigella and Rhizobium) for all my tests during the development of my thesis. 

------

### WHAT I HAVE DONE:

1. I have created three images of each genome when I combined them in the same PCA (This data is not normalized)

2. I have sent to the cluster a file that runs `metaMDS` from the vegan library [Scripts to sent this job](/Scripts/submit_metaMDS_in_cluster.R)

3. I know that I should not use Colsums as a normalized method

------

### WHAT I HAVE TO DO:

1. I have to create the new three images with normalized data (Z score, sqrt)

2. I have to review the literature about k-mers frequency in Prokaryota and the OligonucleotideFrequency function of BioStrings,which will allow to get better counts

3. I have to test the NMS and metric scaling methods

4. I have to search what statistical methods will allow to know which variables affect more my 3 principal components for
the color vector

------

### IMAGES FROM PCA WITH THE THREE GENOMES COMBINED:

A. I have fragmented the genomes with a Kmer of 5000 and overlapping of 3000. Then, I got a matrix of number of fragments vs 256 columns (combinations of possible tetranucleotides (4^4)). After that, I got the matrices for each genome and mixed them in a big matrix.

B. Then, I executed a PCA with prcomp() function in R and got a matrix of 8831 rows and 3 columns (First three components).

C. I got the main 3 components for each genome because I knew how many fragments each genome had at the beginnig => E. coli (2678 fragments) , Shigella (2440 fragments) and Rhizobium (3713 fragments).

D. Later, I created the hex color vector for each genome to be able to create the image independently. 

E. Finally, I draw the image for each genome.
  
===================================================================================================

### E. coli => (26 rows by 103 columns)
![center](/Figs/Semana3/Ecoli_last.png) 

### Shigella => (20 rows by 122 columns)
![center](/Figs/Semana3/Shigella_last.png)

### Rhizobium => (29 rows by 128 columns)
![center](/Figs/Semana3/Rhizobium_last.png)

===================================================================================================

### CHANGES IN: Function3 and Function4

### Function 3:

    Getting_first_three_components_MODI<-function(matrix_for_PCA){
    
    ######## PCA with prcomp#########
    
    prcomp_All<-prcomp(matrix_for_PCA)
    
    #obtaing the sum of varianza of the first three components
    
    Var<-prcomp_All$sdev^2 / sum(prcomp_All$sdev^2)
    
    Variance_all_3<-Var[1:3]
    
    Suma_total<-sum((Variance_all_3)*100)
    
    ## obteniendo eigen of first three components 
    
    loadings_prcomp<-prcomp_All$x
    
    Main_3_Comps<-loadings_prcomp[,c(1,2,3)]
    
    #dim(loadings_prcomp)
    
    return(list("Var_C1_Red"=Variance_all_3[1],"Var_C2_Green"=Variance_all_3[2],
                "Var_C3_Blue"=Variance_all_3[3],"Total_First_3_Comps"=Suma_total,
                "Main_3_Comps"=Main_3_Comps,"pca"=prcomp_All))
    
    }

### Function 4:

    Generating_Hex_Vector_MODI<-function(Main_3_Comps){
    
    # getting min and max for the vector of each color
    
    minAll<-apply(Main_3_Comps,2,min)
    maxAll<-apply(Main_3_Comps,2,max)
    #   
    #   Main_3_Comps
    
    ## getting ranges => 1=Red, 2=Blue, 3=Green
    range_2_color_RED<-c(minAll[1],maxAll[1])
    
    range_2_color_GREEN<-c(minAll[2],maxAll[2])
    
    range_2_color_BLUE<-c(minAll[3],maxAll[3])
    
    range_RGB_color<-c(0,1)
    
    #   #making linear regression
    #   lm_all.out<-c(lm(range_RGB_color~range_2_color_RED),
    #                 lm(range_RGB_color~range_2_color_GREEN),
    #                 lm(range_RGB_color~range_2_color_BLUE))
    
    lm_RED<-lm(range_RGB_color~range_2_color_RED)
    lm_GREEN<-lm(range_RGB_color~range_2_color_GREEN)
    lm_BLUE<-lm(range_RGB_color~range_2_color_BLUE)
    
    #getting slope and intercept
    slope_all<-c(lm_RED$coefficients[2],
                 lm_GREEN$coefficients[2],
                 lm_BLUE$coefficients[2])
    
    intercept_all<-c(lm_RED$coefficients[1],
                     lm_GREEN$coefficients[1],
                     lm_BLUE$coefficients[1])
    
    #normalizing pca results to RGB for each Component (RED,GREEN, and BLUE)
    
    
    mRed<-(Main_3_Comps[,1]*slope_all[1])+intercept_all[1]
    
    mGreen<-(Main_3_Comps[,2]*slope_all[2])+intercept_all[2]
    
    mBlue<-(Main_3_Comps[,3]*slope_all[3])+intercept_all[3]
    
    new_Matriz<-cbind(mRed,mGreen,mBlue)
    
    #if there are negative values => put a zero in their positions
    
    new_Matriz[new_Matriz<0]<-0
    
    # Generating Hex Color Codes
    
    Hex_Color_Vector<-vector()
    
    row_Final<-length(new_Matriz[,1])
    
    columns_Final<-length(new_Matriz[1,])
    
    for(i in 1:row_Final){
      
      for(j in 1:columns_Final){
        
        red<-new_Matriz[i,1]
        green<-new_Matriz[i,2]
        blue<-new_Matriz[i,3]
        
        Hex_Color_Vector[i]<-rgb(red,green,blue,maxColorValue = 1)
        
      }
      
    }
    
    return(Hex_Color_Vector)
    
    
    }
