---
layout: post
title: Improvement on Second week
description: ""
category: Thesis
tags: [thesis, other images]
---
{% include JB/setup %}

### Second Results

During these two weeks, I have been studying about multivariate statistics according to the ```GUide to STasitical Analysis in Microbial Ecology (GUSTA ME)``` [online guide](https://sites.google.com/site/mb3gustame/).

This guide helped to understand what R packages and data, I should use in order to get a better analysis. Also, I have improved the source code of different functions with the purpose of having better results.

===================================================================================================

### Different Images of Ecoli Genome (10711 fragments that are 500bp long)

===================================================================================================

### Image1 => Ecoli from non-normalized input data (Doing PCA)
![center](/Figs/Semana1/Ecoli_Non_normalized_input_for_PCA.png) 

### Image2 => Ecoli from normalized input data by Max value (Doing PCA)
![center](/Figs/Semana1/Ecoli_for_PCA_with_normalized_data_by_Max.png)

### Image3 => Ecoli from normalized input data by Colsums (Doing PCA)
![center](/Figs/Semana1/Ecoli_for_PCA_with_normalized_data_by_ColSums.png)

### Image4 => Ecoli from non-normalized input data of cmdscale$points (Classic MDS)
![center](/Figs/Ecoli_Points_from_cmdscale_non_normalized_test1.png)

==============================================================================================================

## To-Do

1. `I have to see what other transformations` are possible *(For this week)*
2. `I have to do a PCo and Non-metric Multidimensional Scaling (NMDS)` => vegan. ```(Doing right now)```
3. `I have improved some functions` but I still have to see what are the more reliable variables *(For this week)*
4. `I still have to evaluate others lengths of fragments (e.g. 1000, 5000, etc)` and slide through
the genomes with overlaping windows ```(I haven't done it)```

==============================================================================================================
 

### Function1 => Processing_fragments(PATH_FILES)

    Processing_fragments<-function(PATH_FILES){
    
    #GETTING THE FILES AFTER FRAGMENTS OF 500
    files <-list.files(path=PATH_FILES, pattern=".fna500mer", full.names=T, recursive=FALSE)
    
    #GETTING THE FILES READING AS FASTA
    ncrna<-lapply(files, function(x) { read.fasta(x,as.string = T,seqonly = T) })
    
  
    ##GETTING ALL THE FRAGMENTS
    
    fragments_length_list<-vector()
    
    ## Loop to get the number of fragments for each file
    for(i in 1:length(files))
    {
      n<-length(ncrna[[i]])
      fragments_length_list[i]<-as.numeric(n)
      
    }
  
    ## APPENDING EACH TEMP LIST OF FRAGMENTS IF ALLFRAGMENTS
    
    allFragmentsBetter<-list()
    
    for(j in 1:length(ncrna))
    {
      inicio = 1
      final = fragments_length_list[j]
      
      #cat(inicio,"\n",final,"\n")
    
      temp_genome_list<-list()
      
      for(k in 1:final)
      {
        temp_genome_list[k]<-ncrna[[j]][[k]]     
      }
      
      allFragmentsBetter<-append(allFragmentsBetter,temp_genome_list)
      
    }
    
    return(allFragmentsBetter)
    }
  
### Function2 => Getting_frequency_account(allFragments,kmer)

    Getting_frequency_account<-function(allFragments,kmer){
    
    #CONVERTING LOS FRAGMENTOS DE CADA FILE A OBJETOS DE DNAString
    
    DNA_String_Set_list_ALL<-list()
    
    for(i in seq_along(allFragments))
    {
      DNA_String_Set_list_ALL[i]<-DNAStringSet(allFragments[[i]])
    }
    
    # counting oligonucleotide
    countGenome1_Tetra<-lapply(DNA_String_Set_list_ALL,function(x) {oligonucleotideFrequency((x),kmer, as.prob = T) })
    
    # MATRIX FOR THE PCA
    
    #names columns
    col_names<-dimnames(countGenome1_Tetra[[1]])
    col_names<-col_names[[2]]
    
    #names rows
    frag_names<-c(paste("frag",c(1:length(allFragments)),sep=""))
    
    #matrix for PCA
    matrix_PCA<-matrix(unlist(countGenome1_Tetra),nrow = length(allFragments),ncol=256,byrow = T,dimnames=list(frag_names,col_names))
    
    return(matrix_PCA)
    
    }

### Function3 => Getting_first_three_components(matrix_for_PCA_or_MDS)

    Getting_first_three_components<-function(matrix_for_PCA){
  
    ######## PCA with prcomp#########
    
    prcomp_All<-prcomp(matrix_for_PCA)
    
    #obtaing the sum of varianza of the first three components
    
    Var<-prcomp_All$sdev^2 / sum(prcomp_All$sdev^2)
    
    Varianza_3_first_comp<-Var[1:3]
    
    Varianza_3_first_comp_Porcent<-Varianza_3_first_comp*100
    
    Suma_total<-sum(Varianza_3_first_comp_Porcent)
    
    ## obteniendo eigen of first three components 
    
    loadings_prcomp<-prcomp_All$x
  
    #dim(loadings_prcomp)
    
    First_three_components<-loadings_prcomp[,c(1,2,3)]
    
    return(list("Components3"=First_three_components,"VarComponents3"=Suma_total))
  
    }

### Function4 => Generating_hex_color_codes(First_three_components)

      Generating_hex_color_codes<-function(First_three_components){
      
      # getting min and max
      min<-min(First_three_components)
      max<-max(First_three_components)
      
      # getting ranges
      range_2_color<-c(min,max)
      range_RGB_color<-c(0,1)
      
      #making linear regression
      lm.out<-lm(range_RGB_color~range_2_color)
      
      #getting slope and intercept
      slope<-lm.out$coefficients[2]
      intercept<-lm.out$coefficients[1]
      
      #normalizing pca results to RGB
      new_Matriz<-(First_three_components*slope)+intercept
      
      #using funcion rgb to generate matrix of hex color code
      
      # list de cada r,g,b de cada fragmento
      
      Hex_Color_Vector<-vector()
      
      row_Final<-length(new_Matriz[,1])
      
      columns_Final<-length(new_Matriz[1,])
      
      for(i in 1:row_Final){
        
        for(j in 1:columns_Final){
          
          red<-new_Matriz[i,1]
          green<-new_Matriz[i,2]
          blue<-new_Matriz[i,3]
          
          Hex_Color_Vector[i]<-rgb(red,green,blue,maxColorValue = 1)
          
        }
        
      }
      
      return(Hex_Color_Vector)
      
    }

### Function5 => Obtaining_dimensions_for_image(Hex_color_Vector)

    Obtaining_dimensions_for_image<-function(Hex_color_Vector){
      
      len<-length(Hex_color_Vector)
      
      number_rows<-0
      number_cols<-0
      dimensions_vector<-vector()
      
      #first I have to know if len is odd or even
      
      if(len%%2==0){
        
        factors_of_hex_vector<-FUN(len)
        
        positive_factors<-factors_of_hex_vector$pos[factors_of_hex_vector$pos>=100]
        
        number_rows<-as.integer(positive_factors[1])
        
        number_cols<-as.integer(len/number_rows)
        
      }
      else{
        
        len<-len-1
        
        factors_of_hex_vector<-FUN(len)
        
        positive_factors<-factors_of_hex_vector$pos[factors_of_hex_vector$pos>=100]
        
        number_rows<-as.integer(positive_factors[1])
        
        number_cols<-as.integer(len/number_rows)
      }
      
      #dimensions_vector<-c(number_rows,number_cols,len)
      
      return(list("nuRows"=number_rows,"nuCols"=number_cols,"lenForImage"=len))
    }

### Function6 => Creating_image(nu_rows,nu_cols,...)

    library(grid)
    
    Creating_image<-function(nu_rows,nu_cols,len_for_color_matriz,Hex_color_vector){
      
      grid_matrix<-matrix(unlist(Hex_color_vector[1:len_for_color_matriz]), 
                          nrow=nu_rows, ncol = nu_cols, byrow = TRUE)
      
      
      imageFinal<-grid.raster(grid_matrix, interpolate=FALSE)
      
      return(imageFinal)
      
    }
    
    FUN <- function(x) {
      x <- as.integer(x)
      div <- seq_len(abs(x))
      factors <- div[x %% div == 0L]
      factors <- list(neg = -factors, pos = factors)
      return(factors)
    }

