---
layout: post
title: Improvement on Week 8
description: "New advances"
category: Thesis
tags: [thesis, studying data]
---

{% include JB/setup %}

#### Important Note: If you get lost reading this post, you can go to the [Introduction post](http://kamynz.github.io/thesis/2015/07/30/Introduction-of-Thesis/) to understant the objective of my thesis

------

# New Advances

I have been working in the pre-processing of the raw data and looking for the parameters that are giving the best results, so I
want to specify some key points about the methodology of my thesis:

1. Downloading genomes from each domain (e.g. Archaea, Bacteria and Eukaryota)

    * I have 62 genomes so far
    * I am going to select the biggest chromosomes for each Eukaryota

2. Fragmenting genomes with Pyfasta package in order to get 2025 fragments for each genome

    * I am doing images from overlapping fragments (50% of overlapping) and non-overlapping fragments

3. Calcuting tetranucleotide frequencies and biases for each genome in order to make a matrix like the following:

    * Frequency matrix (Image)

4. Using PCA, PCoA and NMDS in order to get the first 3 components

5. Generating squared image in which nrows=45 and ncols=45

------

## Last Images from 7 Genomes non-overlapping fragments

(E. coli, Shigella, Rhizobium, Saccharomyces (Chromoses 15 and 7), Drosophila (Chromoses Sexual and id:3777)

------------

### ``E. coli``

![center](/Figs/Semana4/Ecoli/Ecoli_MardiaTest_qqplot.png)

### ``Shigella``

![center](/Figs/Semana4/Ecoli/Ecoli_MardiaTest_qqplot.png)


### ``Rhizobium``

![center](/Figs/Semana4/Ecoli/Ecoli_MardiaTest_qqplot.png)

### ``Saca 15``

![center](/Figs/Semana4/Ecoli/Ecoli_MardiaTest_qqplot.png)

### ``Saca 7``

![center](/Figs/Semana4/Ecoli/Ecoli_MardiaTest_qqplot.png)

### ``Drosophila Sexual``

![center](/Figs/Semana4/Ecoli/Ecoli_MardiaTest_qqplot.png)

### ``Drosophila id:3777``

![center](/Figs/Semana4/Ecoli/Ecoli_MardiaTest_qqplot.png)








------
