---
layout: post
title: Improvement on Third week
description: "New advances"
category: Thesis
tags: [thesis, new images]
---

{% include JB/setup %}

### Third Results

During this week, I have discussed my results with members of the BCEM [web page](http://bcem.uniandes.edu.co/wordpress/). Then, I decided to use three genomes (E. coli, Shigella and Rhizobium) for all my tests during the development of my thesis. 

### WHAT I HAVE DONE:

1. I have created three images of each genome when I combined them in the same PCA (This data is not normalized)

2. I have sent to the cluster a file that runs `metaMDS` from the vegan library [Scripts to sent this job](kamynz.github.com/Scripts/)

3. I know that I should not use Colsums as a normalized method

### WHAT I HAVE TO DO:

1. I have to create the new three images with normalized data (Z score, sqrt)

2. I have to review the literature about k-mers frequency in Prokaryota and the OligonucleotideFrequency function of BioStrings,which will allow to get better counts

3. I have to test the NMS and metric scaling methods

4. I have to search what statistical methods will allow to know which variables affect more my 3 principal components for
the color vector

### IMAGES FROM PCA WITH THE THREE GENOMES COMBINED:

a. I have fragmented the genomes with a Kmer of 5000 and overlapping of 3000. Then, I got a matrix of number of fragments vs 256 columns (combinations of possible tetranucleotides (4^4)). After that, I got the matrices for each genome and mixed them in a big matrix

b. Then, I executed a PCA with prcomp() function in R and got a matrix of 8831 rows and 3 columns (First three components)

c. I got the main 3 components for each genome because I knew how many fragments each genome had at the beginnig => E. coli (2678 fragments) , Shigella (2440 fragments) and Rhizobium (3713 fragments)

d. Later, I created the hex color vector for each genome to be able to create the image independently. 

e. Finally, I draw the image for each genome
  
===================================================================================================

### E. coli => (26 rows by 103 columns)
![center](/Figs/Semana3/Ecoli_last.png) 

### Shigella => (20 rows by 122 columns)
![center](/Figs/Semana3/Shigella_last.png)

### Rhizobium => (29 rows by 128 columns)
![center](/Figs/Semana3/Rhizobium_last.png)

===================================================================================================

### CHANGES IN: Function3 and Function 4

### Function 3:

  Getting_first_three_components_MODI<-function(matrix_for_PCA){
    
    ######## PCA with prcomp#########
    
    prcomp_All<-prcomp(matrix_for_PCA)
    
    #obtaing the sum of varianza of the first three components
    
    Var<-prcomp_All$sdev^2 / sum(prcomp_All$sdev^2)
    
    Variance_all_3<-Var[1:3]
    
    Suma_total<-sum((Variance_all_3)*100)
    
    ## obteniendo eigen of first three components 
    
    loadings_prcomp<-prcomp_All$x
    
    Main_3_Comps<-loadings_prcomp[,c(1,2,3)]
    
    #dim(loadings_prcomp)
    
    return(list("Var_C1_Red"=Variance_all_3[1],"Var_C2_Green"=Variance_all_3[2],
                "Var_C3_Blue"=Variance_all_3[3],"Total_First_3_Comps"=Suma_total,
                "Main_3_Comps"=Main_3_Comps,"pca"=prcomp_All))
    
  }






  

  





